bindings/go/repository/component/resolver/v1/repository.go

```diff
@@ -0,0 +1,306 @@
+			continue // try the next repository
+		}
+		if err != nil {
+			return nil, fmt.Errorf("getting component version %s/%s from repository %v failed: %w", component, version, repo, err)
+		}
+		return desc, nil
+	}
+	return nil, fmt.Errorf("component version %s/%s not found in any repository", component, version)
+}
+
+func (r *ResolverRepository) ListComponentVersions(ctx context.Context, component string) ([]string, error) {
+	repos := r.RepositoriesForComponentIterator(ctx, component)
+
+	var versionsMu sync.Mutex
+	accumulatedVersions := make(map[string]struct{})
+
+	errGroup, ctx := errgroup.WithContext(ctx)
+	errGroup.SetLimit(r.goRoutineLimit)
+
+	for repo, err := range repos {
+		errGroup.Go(func() error {
+			if err != nil {
+				return fmt.Errorf("getting repository for component %s failed: %w", component, err)
+			}
+			var versions []string
+			versions, err = repo.ListComponentVersions(ctx, component)
+			if err != nil {
+				return fmt.Errorf("listing component versions for %s failed: %w", component, err)
+			}
+			if len(versions) == 0 {
+				slog.DebugContext(ctx, "no versions found for component", "component", component, "repository", repo)
+				return nil
+			}
+			slog.DebugContext(ctx, "found versions for component", "component", component, "versions", versions, "repository", repo)
+			versionsMu.Lock()
+			defer versionsMu.Unlock()
+			for _, version := range versions {
+				if r.matchesAnyResolver(component) {
+					accumulatedVersions[version] = struct{}{}
+				}
+			}
+			return nil
+		})
+	}
+
+	if err := errGroup.Wait(); err != nil {
+		return nil, fmt.Errorf("listing component versions for %s failed: %w", component, err)
+	}
+
+	versionList := slices.Collect(maps.Keys(accumulatedVersions))
+	slices.Sort(versionList)
+
+	return versionList, nil
+}
+
+func (r *ResolverRepository) RepositoriesForComponentIterator(ctx context.Context, component string) iter.Seq2[repository.ComponentVersionRepository, error] {
+	return func(yield func(repository.ComponentVersionRepository, error) bool) {
+		r.matcherCacheMu.RLock()
+		defer r.matcherCacheMu.RUnlock()
+		for index, resolver := range r.resolvers {
+			if !r.matcherCache[index].Match(component, "") {
+				continue
+			}
+
+			repo, err := r.getRepositoryFromCache(ctx, index, resolver)
+			if err != nil {
+				yield(nil, fmt.Errorf("getting repository for resolver %v failed: %w", resolver, err))
+				return
+			}
+
+			slog.DebugContext(ctx, "yielding repository for component", "realm", Realm, "component", component, "repository", resolver.Repository)
+			if !yield(repo, nil) {
+				return
+			}
+		}
+	}
+}
+
+func (r *ResolverRepository) GetResolvers() []*resolverspec.Resolver {
+	return deepCopyResolvers(r.resolvers)
+}
+
+func (r *ResolverRepository) matchesAnyResolver(component string) bool {
+	r.matcherCacheMu.RLock()
+	defer r.matcherCacheMu.RUnlock()
+	for _, m := range r.matcherCache {
+		if m.Match(component, "") {
```

Comment:
is this matching safe? wondering if empty string can cause side effects

Reply:
version is not implemented in the binding but prepared in the signature - its `version`

```go
func (m *ResolverMatcher) Match(componentName, version string) bool {
	return m.componentMatcher.Match(componentName)
}
```